### PR TITLE
Edit Site: Add multiple template loading

### PR DESCRIPTION
**Pull request by epiqueras**
Follows #19081

## Description

This PR continues the `edit-site` work by loading all available templates and allowing you to switch between them.

## How has this been tested?

It was verified that all available templates are now shown in a dropdown in the editor sidebar and that it is possible to change between them, edit single or multiple ones, save single or multiple ones, and see the changes on the front end.

## Screenshots

https://youtu.be/__VN9VNtT0E

<a href="http://www.youtube.com/watch?feature=player_embedded&v=__VN9VNtT0E
" target="_blank">Gutenberg Full Site Editing Experiment Site Editor Demo</a>

## Types of Changes

*New Feature:* The site editor now lets you edit all available templates.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/edit-site/src/components/block-editor/index.js [hunk 1]
```diff
@@ -69,6 +90,17 @@ export default function BlockEditor( { settings: _settings } ) {
 			onChange={ setContent }
 		>
 			<BlockEditorKeyboardShortcuts />
+			<Sidebar.TemplatesFill>
+				<TemplateSwitcher
```
Comment (mtias) [inline packages/edit-site/src/components/block-editor/index.js]: Can we move the template switcher to the editor header?

Comment (mtias) [reply packages/edit-site/src/components/block-editor/index.js]: Actually both features, the "add template" as well.

File: packages/edit-site/src/components/save-button/index.js [hunk 2]
```diff
@@ -48,7 +53,7 @@ export default function SaveButton() {
 				isBusy={ isSaving }
 				onClick={ disabled ? undefined : open }
 			>
-				{ __( 'Update' ) }
+				{ __( 'Customize' ) }
```
Comment (mtias) [inline packages/edit-site/src/components/save-button/index.js]: It'd make more sense for this to say "Update Design" or similar :)

**Comment by epiqueras** [comment]
Yes, I plan on doing that. With an improved version of our dropdown menu.

**Comment by mtias** [comment]
Great to see this in conjunction with the entities save modal.

File: packages/edit-site/src/components/block-editor/index.js (hunk 1 above)
Comment (epiqueras) [reply packages/edit-site/src/components/block-editor/index.js]: I wanted to do that, but we'll need to do some work on improving dropdown menus for it. I want to integrate template parts first and get it ready feature-wise for testing.

It'd be nice to develop this in conjunction with the [new design stuff](https://github.com/WordPress/gutenberg/issues/18667). There is a lot more we can do in this screen, incrementally, that we can't do in the post editor. cc @ItsJonQ @pablohoneyhoney @jasmussen 

What do you think?

File: packages/edit-site/src/components/save-button/index.js (hunk 2 above)
Comment (epiqueras) [reply packages/edit-site/src/components/save-button/index.js]: Agreed, updated 😄

File: packages/edit-site/src/components/block-editor/index.js (hunk 1 above)
Comment (mtias) [reply packages/edit-site/src/components/block-editor/index.js]: I think that's separate, we should still be able to use a dropdown menu from the existing components and list the current template.

Comment (epiqueras) [reply packages/edit-site/src/components/block-editor/index.js]: Sure, but it won't look like the mockup of the template switcher that was made.

I guess, it's still a step in the right direction. I'll do it when I add template parts.

**Comment by dashkevych** [comment]
Are we allowed to add custom CSS classes to the blocks in these templates (for example, if the theme has decided to use utility-first CSS framework) ? 

If we are allowed to use custom CSS classes in the templates, how the template updates will work if the template will be saved as CPT after the user's changes.

**Comment by jffng** [comment]
This is so exciting, thanks @epiqueras!

I tested this out and was wondering about two things, certainly with the understanding that this is work in progress :) 

1. All the available templates show in the site editor dropdown, but not the template parts yet. Will this be a part of a separate PR, or am I missing something? 

2. The UI for adding templates from the site editor was confusing to me. Because of its location and persistence in the editor sidebar, it feels like it refers to the template that is currently active, rather than what it is: a persistent menu for adding new templates of a given name. 

Maybe the action for adding a template could be moved to the top area, and use the current "Templates" area for only editing the active templates name / slug. This feels more in line with the mental model of the "Document" settings in the rest of the editor.

https://cl.ly/6311242bf568

**Comment by epiqueras** [comment]
>All the available templates show in the site editor dropdown, but not the template parts yet. Will this be a part of a separate PR, or am I missing something?

Template parts will be managed via the template part block's edit flow. They show in the template switcher only if they are rendered in one of the loaded templates. Think of it like a zoomed-in view that you could also trigger by directly interacting with a template part block.

>The UI for adding templates from the site editor was confusing to me. Because of its location and persistence in the editor sidebar, it feels like it refers to the template that is currently active, rather than what it is: a persistent menu for adding new templates of a given name.

Yes, this is not meant to represent the final design. I just needed to expose a quick way of doing it.

>Maybe the action for adding a template could be moved to the top area, and use the current "Templates" area for only editing the active templates name / slug. This feels more in line with the mental model of the "Document" settings in the rest of the editor.

I think the idea is for it to be an option in the template switcher dropdown.

**Comment by epiqueras** [comment]
>Template parts will be managed via the template part block's edit flow. They show in the template switcher only if they are rendered in one of the loaded templates. Think of it like a zoomed-in view that you could also trigger by directly interacting with a template part block.

@jffng 

https://github.com/WordPress/gutenberg/pull/19203#issuecomment-567201464

File: lib/demo-block-templates/front-page.html [hunk 3]
```diff
@@ -1,3 +1,4 @@
+<!-- wp:template-part {"slug":"header","theme":"twentytwenty"} /-->
```
Comment (mtias) [inline lib/demo-block-templates/front-page.html]: Do we need the theme slug here? It'd be good if it could be implicit to the "current theme" as set in options.

Comment (epiqueras) [reply lib/demo-block-templates/front-page.html]: It's so that customized template parts can persist after switching themes. Different themes could both define headers.

Do you mean to make it default to the current theme if not set? We can do that, but we would need to save it on customized templates so that they don't load a different theme's template parts after a theme switch. It seems like a lot of additional implicit behaviors that might confuse people. I'm not sure saving one attribute in theme files is worth it.

Comment (mtias) [reply lib/demo-block-templates/front-page.html]: It's a weird requirement to have just for the sake of switching themes. It should be possible to handle it at the time of switching themes as well (also depending on what the UX looks like for what you want to preserve and what you want to replace) since we know both the previous and after contexts. Also not even sure that we want the concept of mixed parts between themes, I'd lean more towards a split of just `current theme` and `user templates`.

Comment (epiqueras) [reply lib/demo-block-templates/front-page.html]: >It should be possible to handle it at the time of switching themes as well (also depending on what the UX looks like for what you want to preserve and what you want to replace) since we know both the previous and after contexts.

That is the implicit behavior I mentioned could be confusing.

>Also not even sure that we want the concept of mixed parts between themes, I'd lean more towards a split of just `current theme` and `user templates`.

It would be confusing if, after switching themes, those `user templates` started referencing different template parts, hence why we require the attribute after customization.

We could default to the theme that originally created each `user template`, but then how does the user add new template parts to existing `user templates`.

**Comment by youknowriad** [comment]
I'm using the "demo templates", I've got two errors:

 - When switching from "Front Page" to "index", I have a lot of blocks broken with a JS error about "tinymce".
 - When adding a new template from the dropdown (I named it "post"), I got an empty screen.

File: lib/edit-site-page.php [hunk 4]
```diff
@@ -74,11 +79,53 @@ function gutenberg_edit_site_init( $hook ) {
...
+		}
+		if ( isset( $_wp_current_template_part_ids ) ) {
+			$template_part_ids = $template_part_ids + $_wp_current_template_part_ids;
+		}
+		$_wp_current_template_id        = null;
+		$_wp_current_template_name      = null;
+		$_wp_current_template_content   = null;
+		$_wp_current_template_hierarchy = null;
+		$_wp_current_template_part_ids  = null;
+	}
```
Comment (youknowriad) [inline lib/edit-site-page.php]: It would be cool to extract all the logic here into a utility function `gutenberg_get_available_templates` or something like that. this bootstrap is becoming too complex.

File: packages/edit-site/src/components/theme-exporter/index.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import { useCallback } from '@wordpress/element';
+import { MenuItem } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import getBaseThemeZip from './get-base-theme-zip';
+
+export default function ThemeExporter( { ids, templatePartIds } ) {
```
Comment (youknowriad) [inline packages/edit-site/src/components/theme-exporter/index.js]: This PR is doing too much IMO, can't we keep the exported for a separate PR.

**Comment by youknowriad** [comment]
This PR deserves more eyes, both design and tech. cc @aduth @jorgefilipecosta @jasmussen @MichaelArestad

File: lib/edit-site-page.php (hunk 4 above)
Comment (epiqueras) [reply lib/edit-site-page.php]: Good idea. It might be useful for plugin developers as well.

File: packages/edit-site/src/components/theme-exporter/index.js (hunk 5 above)
Comment (epiqueras) [reply packages/edit-site/src/components/theme-exporter/index.js]: I thought it was ok because it's a stand-alone prototype and doesn't change public APIs.

Comment (youknowriad) [reply packages/edit-site/src/components/theme-exporter/index.js]: The PR is too big and harder to test, review and land. It's not about APIs it's about landing small pieces iteratively.

**Comment by epiqueras** [comment]
>When switching from "Front Page" to "index", I have a lot of blocks broken with a JS error about "tinymce".

The location of `tinymce` on `window` changed recently, and blocks that rely on it broke. This is unrelated to this PR.

>When adding a new template from the dropdown (I named it "post"), I got an empty screen.

Do you have a console message? I can't reproduce this.

File: packages/edit-site/src/components/theme-exporter/index.js (hunk 5 above)
Comment (epiqueras) [reply packages/edit-site/src/components/theme-exporter/index.js]: I'll break this one up once it gets pseudo-approved, but I worry we are falling for a dogmatic requirement of keeping PRs small.

There is a higher up-front cost for review sure, but you save on the overhead of stacking PRs, and complex rebases every time a significant change happens on the master branch, or you finally get a review, and one of the base branches is merged.

In general, I agree that smaller PRs tend to be better. Still, there is a place for PRs like this when introducing an entirely new module that doesn't modify existing code, especially when the set of features introduced are almost all sequential. If they were things that could be merged parallelly, I would see value in a parallel approach. Otherwise, you're increasing the total person-hours to get the same code in with the same scrutiny, or less, if you consider different people could review the various sequential PRs without the best context of the previous and future changes.

Comment (youknowriad) [reply packages/edit-site/src/components/theme-exporter/index.js]: I have to disagree, loading multiple entities, and exporting entities seems like two complete different features to me and are better addressed in separate PRs. (no dogme here)

Comment (epiqueras) [reply packages/edit-site/src/components/theme-exporter/index.js]: I wasn't thinking about specific examples, but you can't test exporting a theme without loading the templates.

**Comment by epiqueras** [comment]
Any chance we can merge this since it only touches `edit-site` and doesn't change any public APIs.

It's accumulating conflicts and will only get worse with global styles work.

**Comment by youknowriad** [comment]
I still think we should split this PR and only limit it to template selection.

File: packages/edit-site/src/components/add-template/index.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+export default function AddTemplate( {
+	ids,
+	onAddTemplateId,
+	onRequestClose,
+	isOpen,
+} ) {
+	const slugs = useSelect(
+		( select ) => {
+			const { getEntityRecord } = select( 'core' );
+			return ids.reduce( ( acc, id ) => {
```
Comment (mtias) [inline packages/edit-site/src/components/add-template/index.js]: What's `acc` here?

**Comment by mtias** [comment]
Yes, let's take the exporter out into a separate one and merge. Nice work!

File: packages/edit-site/src/components/add-template/index.js (hunk 6 above)
Comment (epiqueras) [reply packages/edit-site/src/components/add-template/index.js]: "Accumulator", it's become pretty common.

**Comment by epiqueras** [comment]
I've rebased the PR and extracted the exporter.

File: lib/template-loader.php [hunk 7]
```diff
@@ -81,24 +81,44 @@ function gutenberg_override_query_template( $template, $type, array $templates =
  * @param array $block The root block to start traversing from.
  */
 function create_auto_draft_for_template_part_block( $block ) {
```
Comment (youknowriad) [inline lib/template-loader.php]: Would be cool to avoid the requirement of having a "Post Id" to render a template/post/template part but I believe that's a big refactoring.

File: packages/edit-site/src/components/block-editor/index.js [hunk 8]
```diff
@@ -44,38 +47,45 @@ export default function BlockEditor( { settings: _settings } ) {
...
+						<NavigateToLink
+							{ ...fillProps }
+							templateIds={ settings.templateIds }
+							activeId={ settings.templateId }
+							onActiveIdChange={ setActiveTemplateId }
+						/>
+					),
+					[ settings.templateIds, settings.templateId, setActiveTemplateId ]
+				) }
+			</URLPopover.LinkViewer.Fill>
```
Comment (youknowriad) [inline packages/edit-site/src/components/block-editor/index.js]: I'm having some trouble understanding what this does?

File: packages/edit-site/src/components/editor/index.js [hunk 9]
```diff
@@ -18,33 +24,44 @@ import Header from '../header';
...
 			select( 'core' ).getEntityRecord(
 				'postType',
-				'wp_template',
+				settings.templateType,
 				settings.templateId
 			),
-		[]
+		[ settings.templateType, settings.templateId ]
 	);
+	const context = useMemo( () => ( { settings, setSettings } ), [
```
Comment (youknowriad) [inline packages/edit-site/src/components/editor/index.js]: Should we use a store instead to match the pattern used in all other modules `core/edit-post` `core/editor`...?

File: packages/edit-site/src/components/header/index.js [hunk 10]
```diff
@@ -1,24 +1,66 @@
...
-			</h1>
+			<div className="edit-site-header__toolbar">
+				<TemplateSwitcher
+					ids={ settings.templateIds }
+					templatePartIds={ settings.templatePartIds }
+					activeId={ settings.templateId }
+					isTemplatePart={ settings.templateType === 'wp_template_part' }
+					onActiveIdChange={ setActiveTemplateId }
+					onActiveTemplatePartIdChange={ setActiveTemplatePartId }
+					onAddTemplateId={ addTemplateId }
```
Comment (youknowriad) [inline packages/edit-site/src/components/header/index.js]: Seems to me like most of these props are useless, why not just do the switching in the component itself since you have access to the Context (or store if it's refactored)

File: lib/template-loader.php (hunk 7 above)
Comment (epiqueras) [reply lib/template-loader.php]: How would you do it?

File: packages/edit-site/src/components/block-editor/index.js (hunk 8 above)
Comment (epiqueras) [reply packages/edit-site/src/components/block-editor/index.js]: It lets us add options to the link viewer, here that is the button to open the template it links to, if any.

File: packages/edit-site/src/components/editor/index.js (hunk 9 above)
Comment (epiqueras) [reply packages/edit-site/src/components/editor/index.js]: I think that's a bit overkill right now. We can see if we need it later, but I did it this way to avoid boilerplate.

File: packages/edit-site/src/components/header/index.js (hunk 10 above)
Comment (epiqueras) [reply packages/edit-site/src/components/header/index.js]: Because I didn't want to prematurely tie these components to a specific context when other UIs could use them in the future.

File: lib/template-loader.php (hunk 7 above)
Comment (youknowriad) [reply lib/template-loader.php]: I don't know. Maybe create temporary ids for new entities and swap them with the real ids upon save.

Comment (epiqueras) [reply lib/template-loader.php]: Aren't `auto-drafts` temporary entities?

Comment (youknowriad) [reply lib/template-loader.php]: yes but we have to create them, so backend logic.  What if we didn't need to create them at all and just have the first "save" create them.

Comment (youknowriad) [reply lib/template-loader.php]: Anyway, that's not a requirement at all here, more a random thought.

Comment (epiqueras) [reply lib/template-loader.php]: Yes, it's possible, it would just require a lot more changes and code paths in `core-data`.

File: lib/demo-block-templates/front-page.html (hunk 3 above)
Comment (vindl) [reply lib/demo-block-templates/front-page.html]: I also didn't expect to see the theme slug here, since I interpreted these demo templates as a generic fallback that should work with any given theme.

> It's so that customized template parts can persist after switching themes. Different themes could both define headers.

Having said that, I do agree with this part in general. It just didn't seem necessary to me in this particular case. Would it make sense to persist it only once the demo template part has been modified?

**Comment by vindl** [comment]
A note for future reviewers: the template switcher is no longer in the sidebar as indicated in the video linked in PR summary. It's now located in the editor header:

Screenshot 2020-02-20 at 21 20 10

**Comment by vindl** [comment]
Nice work @epiqueras! A couple of things I noticed while testing:

1. Not sure if it's intended, but the newly created template parts are not showing in the template switcher even if I click on `Update Design` and save them explicitly. The show up only after a full page refresh.
2. Empty template parts are not be saved (e.g. if I want to quickly leave some placeholders in page template)
3. Creating a new template part requires a theme slug to be entered - I'd expect this to default to the current theme, but wondering if there are reasons for not doing that?

File: lib/demo-block-templates/front-page.html (hunk 3 above)
Comment (epiqueras) [reply lib/demo-block-templates/front-page.html]: What if the user modifies the template, without touching the part, switches themes, and then ends up with a different template part reference. There are too many edge cases like this where we lose with nothing to win.

**Comment by epiqueras** [comment]
Thanks for taking a look!

>Not sure if it's intended, but the newly created template parts are not showing in the template switcher even if I click on Update Design and save them explicitly. The show up only after a full page refresh.

Yeah, that needs to be implemented. Probably with a REST API, I didn't want to explode this PR with.

>Empty template parts are not be saved (e.g. if I want to quickly leave some placeholders in page template)

I can save placeholders with no problem. Can you provide more info?

>Creating a new template part requires a theme slug to be entered - I'd expect this to default to the current theme, but wondering if there are reasons for not doing that?

It's not implemented yet. It should be part of the template part block redesign.

File: lib/edit-site-page.php [hunk 11]
```diff
@@ -74,11 +79,53 @@ function gutenberg_edit_site_init( $hook ) {
...
+		'get_tag_template',
+		'get_author_template',
+		'get_date_template',
+		'get_archive_template',
+	);
+	$template_ids      = array();
+	$template_part_ids = array();
+	foreach ( $template_getters as $template_getter ) {
+		call_user_func( $template_getter );
+		apply_filters( 'template_include', null );
```
Comment (vindl) [inline lib/edit-site-page.php]: Should we document this filter according to [WordPress documentation standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters)? It might make this logic easier to grok.

File: lib/template-loader.php [hunk 12]
```diff
@@ -81,24 +81,44 @@ function gutenberg_override_query_template( $template, $type, array $templates =
...
-	if ( 'core/template-part' === $block['blockName'] && ! isset( $block['attrs']['id'] ) ) {
-		$template_part_file_path =
-			get_stylesheet_directory() . '/block-template-parts/' . $block['attrs']['slug'] . '.html';
-		if ( ! file_exists( $template_part_file_path ) ) {
-			return;
+	global $_wp_current_template_part_ids;
+
+	if ( 'core/template-part' === $block['blockName'] ) {
+		if ( ! isset( $block['attrs']['postId'] ) ) {
+			$template_part_file_path =
```
Comment (vindl) [inline lib/template-loader.php]: Do you think it would be worthwhile to extract this to `get_template_part_file_path( $slug )` helper? It might help with readability here due to nesting, and I imagine we might need this resolution logic elsewhere. It can be abstracted further to include both templates and template parts if needed, since it seems that they follow the same pattern?

**Review by vindl — APPROVED** [review]
Left a couple of comments but I think all of that can be resolved in follow-up PRs if needed. This worked nicely for me. ✨

**Comment by vindl** [comment]
> Template parts will be managed via the template part block's edit flow. They show in the template switcher only if they are rendered in one of the loaded templates. Think of it like a zoomed-in view that you could also trigger by directly interacting with a template part block.

It would be nice if these were shown contextually depending on the currently selected template. After a template is selected these could be expanded below it, somewhat similar to block navigation.

**Comment by vindl** [comment]
> I can save placeholders with no problem. Can you provide more info?

Sure. If I create a new template and add template parts with no content to it (it can be reproduced with post content and post title placeholders too), it shows up fine in the template switcher. After clicking on `Update Design` I only get the option to save the template, and not the added placeholder template parts. After updating and a refresh the newly created template and template parts down show up in the template switcher.

File: lib/edit-site-page.php [hunk 13]
```diff
@@ -28,7 +28,12 @@ class="edit-site"
  * @param string $hook Page.
  */
 function gutenberg_edit_site_init( $hook ) {
-	global $_wp_current_template_id;
+	global
+		$_wp_current_template_id,
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply lib/edit-site-page.php]: It's how we hijack template resolution to resolve block templates.

File: lib/edit-site-page.php [hunk 14]
```diff
@@ -74,11 +79,53 @@ function gutenberg_edit_site_init( $hook ) {
 		$settings['fontSizes'] = $font_sizes;
 	}
 
-	// Get root template by trigerring `./template-loader.php`'s logic.
+	// Get all templates by triggering `./template-loader.php`'s logic.
+	$template_getters  = array(
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply lib/edit-site-page.php]: This needs to match the code from Core template resolution; I think it would be premature to abstract it like that right now.

File: lib/edit-site-page.php (hunk 11 above)
Comment (epiqueras) [reply lib/edit-site-page.php]: Sure

Comment (epiqueras) [reply lib/edit-site-page.php]: But we are not following that convention in the rest of the plugin, are we?

File: lib/template-loader.php (hunk 12 above)
Comment (epiqueras) [reply lib/template-loader.php]: The demo block template code is temporary until we have a proper block-based theme; this will flatten once we can remove it.

**Comment by epiqueras** [comment]
>It would be nice if these were shown contextually depending on the currently selected template. After a template is selected these could be expanded below it, somewhat similar to block navigation.

I think that was mocked up before, but it was discarded because you would be duplicating the entry point to focusing on a template part in all the templates that use it, and it would be harder to grasp all your template parts at once.

>Sure. If I create a new template and add template parts with no content to it (it can be reproduced with post content and post title placeholders too), it shows up fine in the template switcher. After clicking on Update Design I only get the option to save the template, and not the added placeholder template parts. After updating and a refresh the newly created template and template parts down show up in the template switcher.

Yes, that's expected. The template will refresh, and the placeholder will remain, but there is no actual template part associated with the placeholder yet; it's just that, a placeholder for a future one.

File: lib/edit-site-page.php (hunk 11 above)
Comment (vindl) [reply lib/edit-site-page.php]: Now that you mention it, I actually don't remember seeing it around.

File: lib/edit-site-page.php (hunk 4 above)
Comment (youknowriad) [reply lib/edit-site-page.php]: Should we do this as a follow-up?

File: packages/edit-site/src/components/block-editor/index.js (hunk 8 above)
Comment (youknowriad) [reply packages/edit-site/src/components/block-editor/index.js]: Still don't get it? How can we test to understand it more?

File: lib/edit-site-page.php (hunk 4 above)
Comment (epiqueras) [reply lib/edit-site-page.php]: Yes

File: packages/edit-site/src/components/editor/index.js (hunk 9 above)
Comment (youknowriad) [reply packages/edit-site/src/components/editor/index.js]: I'm afraid if we don't do it properly now, we'll forget about it and keep adding adhoc state. I feel we should do it sonner rather than alter.

File: packages/edit-site/src/components/header/index.js (hunk 10 above)
Comment (youknowriad) [reply packages/edit-site/src/components/header/index.js]: It's a bit different pattern from the other components render right after it. I like the fact that the header becomes more about composing things together and keepnig the logic where it belongs (template switching in TemplateSwitcher component). 

I don't really see us reusing that component differently in other places.

File: packages/edit-site/src/components/block-editor/index.js (hunk 8 above)
Comment (epiqueras) [reply packages/edit-site/src/components/block-editor/index.js]: Click on an internal link and see the button to edit the template it links to.

Comment (youknowriad) [reply packages/edit-site/src/components/block-editor/index.js]: I'm having trouble getting an example wiith this. That said, I'm not really confident about the new API (fill) added to achieve this. I also feel this should have deserved a dedicated PR in order to test it, discuss it properly and come up with the best solution.

Is this a feature that was asked for in FSE designs?

Comment (epiqueras) [reply packages/edit-site/src/components/block-editor/index.js]: Ahh, you're having trouble because links are now refactored to use the new `LinkControl`, so this `URLPopover` is not being used anymore. Shouldn't we deprecate it?

>Is this a feature that was asked for in FSE designs?

Yes

File: packages/edit-site/package.json [hunk 15]
```diff
@@ -32,7 +32,9 @@
 		"@wordpress/hooks": "file:../hooks",
 		"@wordpress/i18n": "file:../i18n",
 		"@wordpress/media-utils": "file:../media-utils",
-		"@wordpress/notices": "file:../notices"
+		"@wordpress/notices": "file:../notices",
+		"file-saver": "^2.0.2",
+		"jszip": "^3.2.2"
```
Comment (aduth) [inline packages/edit-site/package.json]: Is this even used? It drew my attention because it's a [26kb+ dependency](https://bundlephobia.com/result?p=jszip@3.2.2), but I don't see that we import it anywhere.

Comment (epiqueras) [reply packages/edit-site/package.json]: It was used to export a theme in a zip file, but that functionality was removed. I think we can keep it as it will soon be introduced again.
